### DISCUSSION TITLE
How to translate the yacc rule into pest grammar

### DISCUSSION THREAD

**Discussion by jbmwdzg** [pest-parser/pest#1022] [Q&A] [answered]
I am working on the C pest grammar development by referring to the ANSI C Yacc grammar. Just have a question 
For this rule in yacc
**postfix_expression
	: primary_expression
	| postfix_expression '[' expression ']'
	| postfix_expression '(' ')'
	| postfix_expression '(' argument_expression_list ')'
	| postfix_expression '.' IDENTIFIER
	| postfix_expression PTR_OP IDENTIFIER
	| postfix_expression INC_OP
	| postfix_expression DEC_OP**

How can I translate into pest, the difficulty is how to resolve the left recursive?

Really appreciate!

**Comment by jbmwdzg** [comment]
It is very important for me to move forward, anyone can share the idea? thanks

**Comment by tomtau** [comment]
How about this: 

```
postfix_expression = { primary_expression ~ postfix_expression_part* }
postfix_expression_part = _{
    "[" ~ expression ~ "]"
    | "(" ~ ")" 
    | "(" ~ argument_expression_list ~ ")"
    | "." ~ IDENTIFIER 
    | PTR_OP ~ IDENTIFIER
    | INC_OP
    | DEC_OP
}
```

**Reply by jbmwdzg** [reply, marked as answer]
@tomtau  Really thanks. I will validate it

**Reply by jbmwdzg** [reply]
@tomtau  One more case, would you kindly advise as well? 
**direct_abstract_declarator
	: '(' abstract_declarator ')'
	| '[' ']'
	| '[' constant_expression ']'
	| direct_abstract_declarator '[' ']'
	| direct_abstract_declarator '[' constant_expression ']'
	| '(' ')'
	| '(' parameter_type_list ')'
	| direct_abstract_declarator '(' ')'
	| direct_abstract_declarator '(' parameter_type_list ')'
	;**

**Reply by jbmwdzg** [reply]
I follow your idea to write down the rule as below

direct_abstract_declarator = {
    ("(" ~ abstract_declarator ~ ")" | "[" ~ "]" | "[" ~ constant_expression ~ "]" | "(" ~ ")" | "(" ~ parameter_type_list ~ ")") ~ direct_abstract_declarator_part*
}

direct_abstract_declarator_part = _{
    "[" ~ "]"
  | "[" ~ "constant_expression " ~ "]"
  | "(" ~ ")"
  | "(" ~ parameter_type_list ~ ")"
}
Still feel something wrong about it, do you have any comments? thanks

**Reply by tomtau** [reply]
1) the way you’ve written the rule, you expect the direct_abstract_declarator to start with a certain pattern (like "(" ~ abstract_declarator ~ ")" or "[" ~ "]" etc.), and then follow with zero or more direct_abstract_declarator_part. But in the direct_abstract_declarator_part, you repeated some patterns that have already been considered in the direct_abstract_declarator.

2) you’ve enclosed "constant_expression " within quotes, which makes it a string literal. I assume constant_expression is another rule in your grammar?

**Reply by jbmwdzg** [reply]
> 1. the way you’ve written the rule, you expect the direct_abstract_declarator to start with a certain pattern (like "(" ~ abstract_declarator ~ ")" or "[" ~ "]" etc.), and then follow with zero or more direct_abstract_declarator_part. But in the direct_abstract_declarator_part, you repeated some patterns that have already been considered in the direct_abstract_declarator.
> 2. you’ve enclosed "constant_expression " within quotes, which makes it a string literal. I assume constant_expression is another rule in your grammar?


Thanks @tomtau  for valuable comments.
For point 1, may I understand that if I remove the duplicated part from direct_abstract_declarator_part  rule, for example, "[" ~ "]", which I have defined in the direct_abstract_declarator , should it accomplish the same purpose as the one in yacc grammar, rt?

For point 2, you are right, it is my mistake, constant_expression  is a rule

**Comment by jbmwdzg** [comment]
I re-defined the rule as below

**direct_abstract_declarator = {
    "(" ~ abstract_declarator ~ ")" ~ direct_abstract_declarator_part*
}

direct_abstract_declarator_part = _{
    "[" ~ "]"
  | "[" ~ constant_expression ~ "]"
  | "(" ~ ")"
  | "(" ~ parameter_type_list ~ ")"
}**

Hope I have correctly understood your comments, ^o^
